### PR TITLE
Add new Shards Capacity Health Indicator

### PR DESCRIPTION
**Pull request by HiDAl**
Introduces a new Health Indicator to check the cluster's health from the shards' capacity perspective. 

It calculates the amount of available room for `data` and `frozen` groups, according to the following rules:

```
if data or frozen nodes have less than 5 shards -> RED
if data or frozen nodes have less than 10 shards -> YELLOW
otherwise -> GREEN
```

This is the output in case the cluster is unhealthy:

```json
GET _health_report/shards_capacity
{
  "cluster_name": "runTask",
  "indicators": {
    "shards_capacity": {
      "status": "red",
      "symptom": "Cluster is close to reaching the configured maximum number of shards for data nodes.",
      "details": {
        "data": {
          "max_shards_in_cluster": 14,
          "current_used_shards": 10
        },
        "frozen": {
          "max_shards_in_cluster": 10,
          "current_used_shards": 0
        }
      },
      "impacts": [
        {
          "id": "elasticsearch:health:shards_capacity:impact:upgrade_blocked",
          "severity": 1,
          "description": "The cluster has too many used shards to be able to upgrade.",
          "impact_areas": [
            "deployment_management"
          ]
        },
        {
          "id": "elasticsearch:health:shards_capacity:impact:creation_of_new_indices_blocked",
          "severity": 1,
          "description": "The cluster is running low on room to add new shards. Adding data to new indices is at risk",
          "impact_areas": [
            "ingest"
          ]
        }
      ],
      "diagnosis": [
        {
          "id": "elasticsearch:health:shards_capacity:diagnosis:increase_max_shards_per_node",
          "cause": "Elasticsearch is about to reach the maximum number of shards it can host, based on your current settings.",
          "action": "Increase the value of [cluster.max_shards_per_node] cluster setting or remove data indices to clear up resources.",
          "help_url": "https://ela.st/fix-shards-capacity"
        }
      ]
    }
  }
}
```

relates #94079 and #91119

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+ * * GREEN otherwise
+ * </p>
+ *
+ *  Although the `max_shard_per_node(.frozen)?` information is scoped by Node, we use the information from master because there is where
+ *  the available room for new shards is checked before creating new indices.
+ */
+public class ShardLimitsHealthIndicatorService implements HealthIndicatorService {
+
+    private static final String NAME = "shard_limits";
+    private final String UPGRADE_BLOCKED = "The cluster has too many used shards to be able to upgrade";
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java:44]: If 10 shards (YELLOW status) are still available - is the upgrade blocked already? 
I guess another way to ask this - is this impact worded correctly for both YELLOW and RED statuses? Or do we need multiple impacts?

Comment (HiDAl) [reply server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java]: mmmh you're right. I did assume that YELLOW is as that bad as RED. Maybe Yellow should be more "hey, __there could be_ issues upgrading in this case", so yes, we'd need 2 impacts.

**Comment by HiDAl** [comment]
@elasticsearchmachine run elasticsearch-ci/part-3

File: server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java [hunk 2]
```diff
@@ -206,16 +206,44 @@ private Result checkShardLimitOnBothGroups(int newShards, int newFrozenShards, C
      * @param state             The cluster state, used to get cluster settings and to get the number of open shards already in the cluster
      */
     public static Result checkShardLimitForNormalNodes(int numberOfNewShards, int replicas, ClusterState state) {
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java:208]: Since we have 2 new methods to check the shard limit could we remove this one (it's only used in one place) and replace its usage with `checkShardLimitForNormalNodes` ?

**Review by andreidan — COMMENTED** [review]
Thanks for working on this Pablo

This generally looks great, I left a few rather minor comments

@tylerperk can you please go through the copy (as most of the output of the API is presented in the UI) 
@shubhaat would you like to have a go through the copy?

File: server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    private static final String UPGRADE_BLOCKED = "The cluster has too many used shards to be able to upgrade";
+    private static final String UPGRADE_AT_RISK = "The cluster is running low on room to add new shards hence upgrade is at risk";
+    private static final String HELP_GUIDE = "https://ela.st/max-shard-limit-reached";
+    static final String NAME = "shard_limits";
+    static final List<HealthIndicatorImpact> RED_INDICATOR_IMPACTS = List.of(
+        new HealthIndicatorImpact(NAME, "upgrade_blocked", 1, UPGRADE_BLOCKED, List.of(ImpactArea.DEPLOYMENT_MANAGEMENT))
+    );
+    static final List<HealthIndicatorImpact> YELLOW_INDICATOR_IMPACTS = List.of(
+        new HealthIndicatorImpact(NAME, "upgrade_at_risk", 2, UPGRADE_AT_RISK, List.of(ImpactArea.DEPLOYMENT_MANAGEMENT))
+    );
```
[the remark this answers is not in the document]
Comment (HiDAl) [reply server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java:53]: that sounds right. I'll add a new impact

**Comment by andreidan** [comment]
@HiDAl the docs needs an update too https://github.com/elastic/elasticsearch/blob/main/docs/reference/health/health.asciidoc

File: server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+ * </p>
+ *
+ *  Although the `max_shard_per_node(.frozen)?` information is scoped by Node, we use the information from master because there is where
+ *  the available room for new shards is checked before creating new indices.
+ */
+public class ShardLimitsHealthIndicatorService implements HealthIndicatorService {
+
+    private static final String UPGRADE_BLOCKED = "The cluster has too many used shards to be able to upgrade";
+    private static final String UPGRADE_AT_RISK = "The cluster is running low on room to add new shards hence upgrade is at risk";
+    private static final String HELP_GUIDE = "https://ela.st/max-shard-limit-reached";
```
[the remark this answers is not in the document]
Comment (HiDAl) [reply server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java:46]: by alias do you mean a short link? this **is** a short link pointing to the settings pages

File: server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        return createIndicator(
+            finalStatus,
+            symptomBuilder.toString(),
+            buildDetails(normalNodes.result, frozenNodes.result),
+            indicatorImpacts,
+            diagnoses
+        );
+    }
+
+    private StatusResult calculateForNormalNodes(HealthMetadata.ShardLimits shardLimits) {
```
[the remark this answers is not in the document]
Comment (HiDAl) [reply server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java:145]: this is my personal opinion, but I don't like mixing static and class methods much, in general. Having a mix of static and class methods becomes quickly messy in tests and code (having to pass a lot of parameters, potential side-effects, etc), from my experience. On the other hand, I think that **only** using side-effect-free static methods makes sense (like in functional programming), but since we already have a framework that ensures that we only have 1 self-contained instance of the object. 

Now, the methods are transitively tested when calling `calculate(...)`. I feel like exposing the methods just for testing it's a bit unneeded since we're not treating the class as a black box.

WDYT?

File: server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java [hunk 6]
```diff
@@ -41,37 +42,35 @@
...
-    private static final BiFunction<String, String, Diagnosis> SHARD_LIMITS_REACHED_FN = (id, settingName) -> new Diagnosis(
+    private static final BiFunction<String, Setting<?>, Diagnosis> SHARD_LIMITS_REACHED_FN = (id, setting) -> new Diagnosis(
         new Diagnosis.Definition(
             NAME,
             id,
-            "The current value of the setting `"
-                + settingName
-                + "` is either about to be or already reached which will not allow adding more shards to the cluster.",
-            "Consider increasing the currently configured value or removing indices to clear up resources " + HELP_GUIDE,
+            "Elasticsearch is about to reach the maximum number of shards it can host.",
```
Comment (tylerperk) [inline server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java:54]: The limit we are about to reach is not a real limit, rather it is based on the setting value. Let's try to find a way to reword this to more closely reflect that. Something like

"Elasticsearch is about to reach the maximum number of shards it can host, based on your current settings."

File: server/src/internalClusterTest/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorServiceIT.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    private void waitForHealthMetadata() throws Exception {
+        assertBusy(() -> {
+            var healthMetadata = HealthMetadata.getFromClusterState(internalCluster().clusterService().state());
+
+            assertNotNull(healthMetadata);
+            assertNotNull(healthMetadata.getShardLimitsMetadata());
+            assertTrue(
+                "max_shards_per_node settings must be greater than 0",
```
Comment (tylerperk) [inline server/src/internalClusterTest/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorServiceIT.java:135]: Should this be settings (plural) or just setting? Same for line 139.

File: server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+            }
+            builder.endObject();
+            return builder;
+        };
+    }
+
+    private HealthIndicatorResult unknownIndicator() {
+        return createIndicator(
+            HealthStatus.UNKNOWN,
+            "No number of shards limits data.",
```
Comment (tylerperk) [inline server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java:202]: If this will be seen by a user (will Cloud UI expose this?) I think we can word it better as I had to read it a few times. 

"No data regarding shard limits was available"? "Unable to determine shard limit status"?

File: server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+ *
+ *  Although the `max_shard_per_node(.frozen)?` information is scoped by Node, we use the information from master because there is where
+ *  the available room for new shards is checked before creating new indices.
+ */
+public class ShardLimitsHealthIndicatorService implements HealthIndicatorService {
+
+    private static final String UPGRADE_BLOCKED = "The cluster has too many used shards to be able to upgrade";
+    private static final String UPGRADE_AT_RISK = "The cluster is running low on room to add new shards hence upgrade is at risk";
+    private static final String HELP_GUIDE = "https://ela.st/max-shard-limit-reached";
+    static final String NAME = "shard_limits";
```
[the remark this answers is not in the document]
Comment (tylerperk) [reply server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java]: I'd need more context - where will a user see this?

[the remark this answers is not in the document]
Comment (HiDAl) [reply server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java]: @tylerperk In this page: Screenshot 2023-03-22 at 12 39 32

[the remark this answers is not in the document]
Comment (HiDAl) [reply server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java]: in the `Deployments > Monitoring > Health` page

File: server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+        return new StatusResult(HealthStatus.GREEN, result);
+    }
+
+    static HealthIndicatorDetails buildDetails(ShardLimitValidator.Result normalNodes, ShardLimitValidator.Result frozenNodes) {
+        return (builder, params) -> {
+            builder.startObject();
+            {
+                builder.startObject(NORMAL_GROUP + "_nodes");
```
[the remark this answers is not in the document]
Comment (HiDAl) [reply server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java:179]: two comments:

1. it might be worth not splitting the details into groups and just using the same format that the settings `cluster.max_shards_per_node & cluster.max_shards_per_node.frozen`, so the details would be like this:
```json
{
  "max_shards_in_cluster": 18,
  "current_used_shards": 10
  "max_shards_in_cluster.frozen": 1,
  "current_used_shards.frozen": 0
}
```

2. wherever we use `normal & frozen` use non-frozen and frozen. Not sure about this, but I can imagine better wording here.

File: server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java:144]: ++ I understand your reasoning.

We generally tend to favour static methods (functions) where we can as they're easier to reason about and they signal they don't use any global state (as you point out, there might be cases where a method would solely rely on the class global state and not only that but, business-wise they make sense to belong to the instance class). 

Testing these methods in isolation becomes much easier as well and the unit tests themselves, when not funnelled through the parent class, start to offer nice documentation around how the functions behave.
e.g. https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleService.java#L250

https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java#L976 

and 

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SnapshotLifecycleServiceTests.java#L388

**Comment by HiDAl** [comment]
@andreidan  I did rename the indicator to `ShardsCapacity`

1. I didn't rename the class `ShardLimitsValidator` because > 30 files are using the class, hence this PR will easily become a mess. I'll rename it in a separate PR.
2. didn't rename the record `ShardLimitsMetadata` because it actually contains the configured limits.

**Comment by andreidan** [comment]
@HiDAl the `ShardLimitsValidator` and `ShardLimitsMetadata` can stay named as they are IMO (they're not user facing and are extensively documented )

Can you please update the PR description to reflect the latest state?

File: server/src/test/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorServiceTests.java [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+                        "frozen",
+                        Map.of("max_shards_in_cluster", 25, "current_used_shards", 22)
+                    )
+                )
+            );
+        }
+    }
+
+    public void testCalculateMethods() {
+        var mockedState = mock(ClusterState.class);
```
Comment (andreidan) [inline server/src/test/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorServiceTests.java:292]: Please replace this with a real ClusterState. This could be `ClusterState.EMPTY_STATE`

File: server/src/test/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorServiceTests.java [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+        XContentParser parser = XContentType.JSON.xContent()
+            .createParser(xContentRegistry(), LoggingDeprecationHandler.INSTANCE, BytesReference.bytes(builder).streamInput());
+        return parser.map();
+    }
+
+    private static ClusterService createClusterService(ClusterState clusterState) {
+        var clusterService = mock(ClusterService.class);
+        when(clusterService.state()).thenReturn(clusterState);
+        return clusterService;
+    }
```
Comment (andreidan) [inline server/src/test/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorServiceTests.java:351]: We prefer to not mock things when we can avoid it (so the test is as close to the real scenario as possible).

Please replace this mock with a real `ClusterService`. 
See
```
org.elasticsearch.test.ClusterServiceUtils.createClusterService
```
and
```
org.elasticsearch.test.ClusterServiceUtils.setState
```

File: server/src/main/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorService.java [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+ *  the available room for new shards is checked before creating new indices.
+ */
+public class ShardsCapacityHealthIndicatorService implements HealthIndicatorService {
+
+    static final String NAME = "shards_capacity";
+
+    static final String DATA_NODE_NAME = "data";
+    static final String FROZEN_NODE_NAME = "frozen";
+    private static final String UPGRADE_BLOCKED = "The cluster has too many used shards to be able to upgrade.";
+    private static final String UPGRADE_AT_RISK = "The cluster is running low on room to add new shards hence upgrade is at risk.";
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorService.java:48]:
```suggestion
    private static final String UPGRADE_AT_RISK = "The cluster is running low on room to add new shard. Upgrading to a new version is at risk.";
```

File: server/src/main/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorService.java [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+    private static final String INDEX_CREATION_BLOCKED =
+        "The cluster is running low on room to add new shards hence the creation of new indices might fail.";
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorService.java:50]:
```suggestion
        "The cluster is running low on room to add new shards. Adding data to new indices is at risk.";
```

File: server/src/main/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorService.java [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    private static final String INDEX_CREATION_RISK =
+        "The cluster is running low on room to add new shards hence the creation of new indices might fail.";
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorService.java:52]:
```suggestion
        "The cluster is running low on room to add new shards. Adding data to new indices might soon fail.";
```

**Review by andreidan — APPROVED** [review]
Thanks for iterating on this Pablo. This LGTM 🚀  - left a few very minor suggestions

File: server/src/test/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorServiceTests.java (hunk 12 above)
Comment (HiDAl) [reply server/src/test/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorServiceTests.java:351]: real nice! I did copy what we're using here https://github.com/elastic/elasticsearch/blob/main/server/src/test/java/org/elasticsearch/health/node/DiskHealthIndicatorServiceTests.java#L1065-L1073

**Comment by HiDAl** [comment]
@andreidan  I've applied all the recommended changes :)

**Comment by HiDAl** [comment]
@elasticsearchmachine run elasticsearch-ci/part-3

**Comment by HiDAl** [comment]
@elasticmachine update branch

**Comment by HiDAl** [comment]
@elasticsearchmachine run elasticsearch-ci/part-1
